zkvm/src/signature/counterparty.rs

```diff
@@ -0,0 +1,109 @@
+use crate::errors::VMError;
+use crate::signature::multikey::Multikey;
+use crate::signature::VerificationKey;
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use subtle::ConstantTimeEq;
+
+#[derive(Copy, Clone)]
+pub struct NoncePrecommitment([u8; 32]);
+
+#[derive(Copy, Clone, Debug)]
+pub struct NonceCommitment(RistrettoPoint);
+
+impl NonceCommitment {
+    pub fn new(commitment: RistrettoPoint) -> Self {
+        NonceCommitment(commitment)
+    }
+
+    pub fn precommit(&self) -> NoncePrecommitment {
+        let mut h = Transcript::new(b"MuSig.nonce-precommit");
+        h.commit_point(b"R", &self.0.compress());
+        let mut precommitment = [0u8; 32];
+        h.challenge_bytes(b"precommitment", &mut precommitment);
+        NoncePrecommitment(precommitment)
+    }
+
+    pub fn sum(commitments: &Vec<Self>) -> RistrettoPoint {
```

Comment:
most rusty (eye-rolling) way to do this is

```suggestion
    pub fn sum<I>(commitments: I) -> RistrettoPoint 
    where 
    I: IntoIterator,
    I::Item: Borrow<RistrettoPoint>
    {
```

and then `commitments.into_iter().map(|p|p.borrow()).sum()`. You can then still pass &Vec into this method since it implements IntoIterator.

Reply:
Is there a benefit (performance, memory-wise) to this approach?
Or is it just for the sake of the rusty eye-rolls? :P

Reply:
if it's going to be a public API, it allows users pass a wider variety of things in. E.g. if they already have an iterator of owned or borrowed points, they can pass it in w/o doing any allocations (perf benefit) or conversions (ergonomics benefit).

Reply:
if it's a private API (maybe it should be - it's only used within Party impl?) - then `&[u8]` is simple and unassuming enough.

Reply:
I actually think that all of the counterparty structs & API should be only usable from within the module, which I denoted by making them all `pub(super)`. So, this should take care of the concern about flexibility in the case of a public API (which it's not)